File: hostel_management_system_backend/Services/HostelImagesService.cs
using AutoMapper;
using hostel_management_system_backend.Exceptions;
using Microsoft.AspNetCore.Http;

public interface IHostelImagesService
{
    Task<List<HostelImageReadDto>> GetImagesByHostelIdAsync(Guid hostelId, CancellationToken cancellationToken);
    Task<HostelImageReadDto> AddImageAsync(Guid hostelId, IFormFile file, int? displayOrder, Guid userId, bool isAdmin, CancellationToken cancellationToken);
    Task DeleteImageAsync(Guid imageId, Guid userId, bool isAdmin, CancellationToken cancellationToken);
    Task UpdateImageOrderAsync(Guid imageId, int order, Guid userId, bool isAdmin, CancellationToken cancellationToken);
}

public sealed class HostelImagesService : IHostelImagesService
{
    private const long MaxImageSizeBytes = 5 * 1024 * 1024;

    private readonly IHostelImageRepository _repository;
    private readonly IImageStorageService _storageService;
    private readonly IMapper _mapper;

    public HostelImagesService(IHostelImageRepository repository, IImageStorageService storageService, IMapper mapper)
    {
        _repository = repository;
        _storageService = storageService;
        _mapper = mapper;
    }

    public async Task<List<HostelImageReadDto>> GetImagesByHostelIdAsync(Guid hostelId, CancellationToken cancellationToken)
    {
        var images = await _repository.GetImagesByHostelIdAsync(hostelId, cancellationToken);
        return _mapper.Map<List<HostelImageReadDto>>(images);
    }

    public async Task<HostelImageReadDto> AddImageAsync(Guid hostelId, IFormFile file, int? displayOrder, Guid userId, bool isAdmin, CancellationToken cancellationToken)
    {
        if (file is null || file.Length <= 0)
        {
            throw new BadRequestException("Image file is required.");
        }

        if (file.Length > MaxImageSizeBytes)
        {
            throw new BadRequestException("Image too large. Maximum allowed size is 5MB.");
        }

        if (string.IsNullOrWhiteSpace(file.ContentType) || !file.ContentType.StartsWith("image/", StringComparison.OrdinalIgnoreCase))
        {
            throw new BadRequestException("Invalid image type.");
        }

        var ownerId = await _repository.GetHostelOwnerIdAsync(hostelId, cancellationToken);
        if (!ownerId.HasValue)
        {
            throw new NotFoundException("Hostel not found.");
        }

        if (!isAdmin && ownerId.Value != userId)
        {
            throw new ForbiddenException("You are not allowed to upload images for this hostel.");
        }

        var storedImage = await _storageService.UploadImageAsync(file, hostelId, cancellationToken);

        try
        {
            var image = new HostelImage
            {
                HostelId = hostelId,
                FileName = storedImage.StoredFileName,
                ContentType = storedImage.ContentType,
                FileSize = storedImage.FileSize,
                ImageUrl = storedImage.ImageUrl,
                DisplayOrder = displayOrder ?? 0,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = null,
                IsDeleted = false
            };

            var created = await _repository.AddImageAsync(image, cancellationToken);
            return _mapper.Map<HostelImageReadDto>(created);
        }
        catch
        {
            await _storageService.DeleteImageAsync(storedImage.ImageUrl, cancellationToken);
            throw;
        }
    }

    public async Task DeleteImageAsync(Guid imageId, Guid userId, bool isAdmin, CancellationToken cancellationToken)
    {
        var image = await _repository.GetImageByIdAsync(imageId, cancellationToken);
        if (image is null)
        {
            throw new NotFoundException("Image not found.");
        }

        if (!isAdmin && image.Hostel.OwnerId != userId)
        {
            throw new ForbiddenException("You are not allowed to delete this image.");
        }

        await _storageService.DeleteImageAsync(image.ImageUrl, cancellationToken);
        await _repository.DeleteImageAsync(image, cancellationToken);
    }

    public async Task UpdateImageOrderAsync(Guid imageId, int order, Guid userId, bool isAdmin, CancellationToken cancellationToken)
    {
        if (order < 0)
        {
            throw new BadRequestException("Display order must be greater than or equal to 0.");
        }

        var image = await _repository.GetImageByIdAsync(imageId, cancellationToken);
        if (image is null)
        {
            throw new NotFoundException("Image not found.");
        }

        if (!isAdmin && image.Hostel.OwnerId != userId)
        {
            throw new ForbiddenException("You are not allowed to reorder this image.");
        }

        await _repository.UpdateImageOrderAsync(image, order, cancellationToken);
    }
}

File: src/hostels/images/hostel-images.service.ts
import { Injectable, Inject } from '@nestjs/common';

import {
  AppBadRequestException,
  AppForbiddenException,
  AppNotFoundException,
} from '../../common/exceptions/app-exception';
import { DatabaseService } from '../../database/database.service';
import { HostelImageReadDto } from './dto/hostel-images.dto';
import { StorageService } from './storage.interface';
import { STORAGE_SERVICE_TOKEN } from './storage.provider';

@Injectable()
export class HostelImagesService {
  constructor(
    private readonly db: DatabaseService,
    @Inject(STORAGE_SERVICE_TOKEN) private readonly storageService: StorageService,
  ) {}

  async getImagesByHostelId(hostelId: string): Promise<HostelImageReadDto[]> {
    const images = await this.db.hostelImage.findMany({
      where: {
        hostelId,
        isDeleted: false,
      },
      orderBy: [{ displayOrder: 'asc' }, { createdAt: 'asc' }],
    });

    return images.map((image) => this.toReadDto(image));
  }

  async addImage(
    hostelId: string,
    file: Express.Multer.File,
    displayOrder: number | undefined,
    userId: string,
    isAdmin: boolean,
  ): Promise<HostelImageReadDto> {
    const hostel = await this.db.hostel.findFirst({
      where: {
        id: hostelId,
        isDeleted: false,
      },
    });

    if (!hostel) {
      throw new AppNotFoundException('Hostel not found.');
    }

    if (!isAdmin && hostel.ownerId !== userId) {
      throw new AppForbiddenException('You are not allowed to upload images for this hostel.');
    }

    const count = await this.db.hostelImage.count({
      where: {
        hostelId,
        isDeleted: false,
      },
    });

    if (count >= 8) {
      throw new AppBadRequestException('Maximum 8 images allowed per hostel');
    }

    const stored = await this.storageService.uploadImage(file, hostelId);

    try {
      const image = await this.db.hostelImage.create({
        data: {
          hostelId,
          fileName: stored.storedFileName,
          contentType: stored.contentType,
          fileSize: BigInt(stored.fileSize),
          imageUrl: stored.imageUrl,
          displayOrder: displayOrder ?? 0,
          createdAt: new Date(),
          isDeleted: false,
        },
      });

      return this.toReadDto(image);
    } catch (error) {
      await this.storageService.deleteImage(stored.imageUrl);
      throw error;
    }
  }

  async deleteImage(imageId: string, userId: string, isAdmin: boolean) {
    const image = await this.db.hostelImage.findFirst({
      where: {
        id: imageId,
        isDeleted: false,
      },
      include: {
        hostel: true,
      },
    });

    if (!image) {
      throw new AppNotFoundException('Image not found.');
    }

    if (!isAdmin && image.hostel.ownerId !== userId) {
      throw new AppForbiddenException('You are not allowed to delete this image.');
    }

    await this.storageService.deleteImage(image.imageUrl);
    await this.db.hostelImage.update({
      where: { id: imageId },
      data: {
        isDeleted: true,
        deletedAt: new Date(),
        updatedAt: new Date(),
      },
    });
  }

  async updateImageOrder(imageId: string, order: number, userId: string, isAdmin: boolean) {
    if (order < 0) {
      throw new AppBadRequestException('Display order must be greater than or equal to 0.');
    }

    const image = await this.db.hostelImage.findFirst({
      where: {
        id: imageId,
        isDeleted: false,
      },
      include: {
        hostel: true,
      },
    });

    if (!image) {
      throw new AppNotFoundException('Image not found.');
    }

    if (!isAdmin && image.hostel.ownerId !== userId) {
      throw new AppForbiddenException('You are not allowed to reorder this image.');
    }

    await this.db.hostelImage.update({
      where: { id: imageId },
      data: {
        displayOrder: order,
        updatedAt: new Date(),
      },
    });
  }

  async deleteImageByUrl(imageUrl: string) {
    await this.storageService.deleteImage(imageUrl);
  }

  private toReadDto(image: any): HostelImageReadDto {
    return {
      id: image.id,
      hostelId: image.hostelId,
      fileName: image.fileName,
      contentType: image.contentType,
      fileSize: Number(image.fileSize),
      imageUrl: image.imageUrl,
      displayOrder: image.displayOrder,
      createdAt: image.createdAt,
      updatedAt: image.updatedAt,
    };
  }
}
